common/djangoapps/student/models.py

```diff
@@ -3116,6 +3120,130 @@ class AccountRecoveryConfiguration(ConfigurationModel):
+    last_day_of_streak = models.DateField(default=None, null=True, blank=True)
+    longest_ever_streak = models.IntegerField(default=0)
+    STREAK_LENGTHS_TO_CELEBRATE = [3]
+    STREAK_BREAK_LENGTH = 1
+
+    def __str__(self):
+        return (
+            '[UserCelebration] user: {}; first_day_of_streak {};  last_day_of_streak {}; longest_ever_streak {};'
+        ).format(self.user.username, self.first_day_of_streak, self.last_day_of_streak, self.longest_ever_streak)
+
+    @classmethod
+    def _get_now(cls):
+        """ Retrieve the value for the current datetime in the user's timezone """
+        # importing here to avoid a circular import
+        from lms.djangoapps.courseware.context_processor import user_timezone_locale_prefs
+        user_timezone_locale = user_timezone_locale_prefs(crum.get_current_request())
+        user_timezone = timezone(user_timezone_locale['user_timezone'] or str(UTC))
+        return user_timezone.localize(datetime.now())
+
+    def _calculate_streak_updates(self):
+        """ Calculate the updates that should be applied to the streak fields of the provided celebration """
+        first_day_of_streak = self.first_day_of_streak
+        last_day_of_streak = self.last_day_of_streak
+        today = self.__class__._get_now().date()  # pylint: disable=protected-access
+        streak_length_to_celebrate = None
+
+        first_ever_streak = None is first_day_of_streak is last_day_of_streak
+        inside_active_streak = last_day_of_streak and first_day_of_streak
+        break_length = timedelta(days=self.STREAK_BREAK_LENGTH)
+        can_start_new_streak = last_day_of_streak and last_day_of_streak + break_length < today
+        already_updated_streak_today = last_day_of_streak == today
+
+        if first_ever_streak or can_start_new_streak:
+            # Start new streak
+            first_day_of_streak = today
+            last_day_of_streak = today
+        elif inside_active_streak:
+            # Increment existing streak
+            last_day_of_streak = today
+
+            streak_length = (last_day_of_streak - first_day_of_streak).days + 1
+            if streak_length in self.STREAK_LENGTHS_TO_CELEBRATE and not already_updated_streak_today:
+                # Celebrate if we didn't already celebrate today
+                streak_length_to_celebrate = streak_length
+        # When the learner is in the middle of a break, we don't take any steps
+
+        return first_day_of_streak, last_day_of_streak, streak_length_to_celebrate
+
+    @classmethod
+    def _update_streak(cls, celebration, first_day_of_streak, last_day_of_streak):
+        """ Update the celebration with the new streak data """
+        # If anything needs to be updated, update the celebration in the database
+        if (first_day_of_streak != celebration.first_day_of_streak or
+                last_day_of_streak != celebration.last_day_of_streak):
```

Comment:
Minor ignorable nits. I'd make this an instance method, not a class method. And I'd do an early exit, rather than indenting the entire method. But that latter especially is a Mike taste thing that you might disagree with.

Reply:
i changed it to an instanced method, agree that that makes sense.  
i kept the if statement as is though because i think it reads clearer to me in this particular case since its closer to how you would say it in english
if one of the variables changed:
updated the celebration
rather than

if the variables are the same 
do nothing

otherwise update the celebration

I do prefer early exits though in cases that are like
if one of the variables is none then exit since we need the variable to be not none for the later code